// ngx-filepond/bundles/ngx-filepond.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('filepond')) :
    typeof define === 'function' && define.amd ? define('ngx-filepond', ['exports', '@angular/core', '@angular/common', 'filepond'], factory) :
    (global = global || self, factory(global['ngx-filepond'] = {}, global.ng.core, global.ng.common, global.filepond));
}(this, (function (exports, core, common, filepond) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __exportStar(m, exports) {
        for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
        if (m) return m.call(o);
        return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    // We test if filepond is supported on the current client
    var isSupported = filepond.supported();
    // List of attributes for fallback input
    var inputAttributes = [
        "id",
        "name",
        "class",
        "multiple",
        "required",
        "disabled",
        "capture",
        "accept",
    ];
    // Methods not made available on the component
    var filteredComponentMethods = [
        "setOptions",
        "on",
        "off",
        "onOnce",
        "appendTo",
        "insertAfter",
        "insertBefore",
        "isAttachedTo",
        "replaceElement",
        "restoreElement",
        "destroy",
    ];
    var outputs = [
        "oninit",
        "onwarning",
        "onerror",
        "onactivatefile",
        "onaddfilestart",
        "onaddfileprogress",
        "onaddfile",
        "onprocessfilestart",
        "onprocessfileprogress",
        "onprocessfileabort",
        "onprocessfilerevert",
        "onprocessfile",
        "onprocessfiles",
        "onremovefile",
        "onpreparefile",
        "onupdatefiles",
    ];
    // Component outline
    var FilePondComponent = /** @class */ (function () {
        function FilePondComponent(root, zone) {
            this.options = {};
            // same as outputs array
            this.oninit = new core.EventEmitter();
            this.onwarning = new core.EventEmitter();
            this.onerror = new core.EventEmitter();
            this.onactivatefile = new core.EventEmitter();
            this.onaddfilestart = new core.EventEmitter();
            this.onaddfileprogress = new core.EventEmitter();
            this.onaddfile = new core.EventEmitter();
            this.onprocessfilestart = new core.EventEmitter();
            this.onprocessfileprogress = new core.EventEmitter();
            this.onprocessfileabort = new core.EventEmitter();
            this.onprocessfilerevert = new core.EventEmitter();
            this.onprocessfile = new core.EventEmitter();
            this.onprocessfiles = new core.EventEmitter();
            this.onremovefile = new core.EventEmitter();
            this.onpreparefile = new core.EventEmitter();
            this.onupdatefiles = new core.EventEmitter();
            this.handleEvent = null;
            this.root = root;
            this.zone = zone;
        }
        FilePondComponent.prototype.ngAfterViewInit = function () {
            var _this = this;
            var input = this.root.nativeElement.querySelector("input");
            // transfer relevant attributes to input, this so we still have an input with the correct attributes should file pond not load
            var attributes = this.root.nativeElement.attributes;
            inputAttributes.forEach(function (name) {
                var value = attributes[name]
                    ? attributes[name].value
                    : _this.options[name];
                if (!value) {
                    return;
                }
                input.setAttribute(name, value);
            });
            // no sufficient features supported in this browser
            if (!isSupported) {
                return;
            }
            // map FilePond events to Angular @outputs
            this.handleEvent = function (e) {
                _this["on" + e.type.split(":")[1]].emit(__assign({}, e.detail));
            };
            outputs.forEach(function (event) {
                _this.root.nativeElement.addEventListener("FilePond:" + event.substr(2), _this.handleEvent);
            });
            // will block angular from listening to events inside the pond
            this.zone.runOutsideAngular(function () {
                // create instance
                _this.pond = filepond.create(input, __assign(__assign({}, _this.options), { 
                    // our initial files
                    files: _this.files }));
            });
            // Copy instance method references to component instance
            Object.keys(this.pond)
                // remove unwanted methods
                .filter(function (key) { return filteredComponentMethods.indexOf(key) === -1; })
                // set method references from the component instance to the pond instance
                .forEach(function (key) {
                _this[key] = _this.pond[key];
            });
        };
        FilePondComponent.prototype.ngOnChanges = function (changes) {
            // no need to handle first change
            if (changes.firstChange) {
                return;
            }
            // no filepond instance available
            if (!this.pond) {
                return;
            }
            // use new options object as base ( or if not available, use current options )
            var options = changes.options
                ? changes.options.currentValue
                : this.options;
            // see if file list has changed
            if (changes.files &&
                JSON.stringify(changes.files.previousValue) !==
                    JSON.stringify(changes.files.currentValue)) {
                // file list has changed
                options.files = changes.files.currentValue;
            }
            // set new options
            this.pond.setOptions(options);
        };
        FilePondComponent.prototype.ngOnDestroy = function () {
            var _this = this;
            if (!this.pond) {
                return;
            }
            outputs.forEach(function (event) {
                _this.root.nativeElement.removeEventListener("FilePond:" + event.substr(2), _this.handleEvent);
            });
            this.pond.destroy();
        };
        FilePondComponent.ctorParameters = function () { return [
            { type: core.ElementRef },
            { type: core.NgZone }
        ]; };
        __decorate([
            core.Input()
        ], FilePondComponent.prototype, "options", void 0);
        __decorate([
            core.Input()
        ], FilePondComponent.prototype, "files", void 0);
        __decorate([
            core.Output()
        ], FilePondComponent.prototype, "oninit", void 0);
        __decorate([
            core.Output()
        ], FilePondComponent.prototype, "onwarning", void 0);
        __decorate([
            core.Output()
        ], FilePondComponent.prototype, "onerror", void 0);
        __decorate([
            core.Output()
        ], FilePondComponent.prototype, "onactivatefile", void 0);
        __decorate([
            core.Output()
        ], FilePondComponent.prototype, "onaddfilestart", void 0);
        __decorate([
            core.Output()
        ], FilePondComponent.prototype, "onaddfileprogress", void 0);
        __decorate([
            core.Output()
        ], FilePondComponent.prototype, "onaddfile", void 0);
        __decorate([
            core.Output()
        ], FilePondComponent.prototype, "onprocessfilestart", void 0);
        __decorate([
            core.Output()
        ], FilePondComponent.prototype, "onprocessfileprogress", void 0);
        __decorate([
            core.Output()
        ], FilePondComponent.prototype, "onprocessfileabort", void 0);
        __decorate([
            core.Output()
        ], FilePondComponent.prototype, "onprocessfilerevert", void 0);
        __decorate([
            core.Output()
        ], FilePondComponent.prototype, "onprocessfile", void 0);
        __decorate([
            core.Output()
        ], FilePondComponent.prototype, "onprocessfiles", void 0);
        __decorate([
            core.Output()
        ], FilePondComponent.prototype, "onremovefile", void 0);
        __decorate([
            core.Output()
        ], FilePondComponent.prototype, "onpreparefile", void 0);
        __decorate([
            core.Output()
        ], FilePondComponent.prototype, "onupdatefiles", void 0);
        FilePondComponent = __decorate([
            core.Component({
                selector: "file-pond",
                encapsulation: core.ViewEncapsulation.None,
                template: "<div class=\"filepond--wrapper\">\n  <input type=\"file\"/>\n</div>",
                styles: [""]
            })
        ], FilePondComponent);
        return FilePondComponent;
    }());

    var FilePondModule = /** @class */ (function () {
        function FilePondModule() {
        }
        FilePondModule = __decorate([
            core.NgModule({
                imports: [
                    common.CommonModule
                ],
                declarations: [FilePondComponent],
                exports: [
                    FilePondComponent
                ]
            })
        ], FilePondModule);
        return FilePondModule;
    }());

    Object.defineProperty(exports, 'registerPlugin', {
        enumerable: true,
        get: function () {
            return filepond.registerPlugin;
        }
    });
    exports.FilePondModule = FilePondModule;
    exports.ɵa = FilePondComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
